<a href="https://colab.research.google.com/github/esther119/Bayesian-Statistics/blob/main/CS146_Session_7_Selecting_prior_hyperparameters.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In this session, we model a data set using the normal likelihood with the normal-inversegamma prior. You will reuse your code from the previous classso that you can calculate
posterior hyperparameters from the prior hyperparameters and data.

● Make sure you type all your solutions in a Jupyter notebook, upload it to Gist, and be ready to paste a link to your work in a class poll.

● Also, have this code ready for a class activity.
In the previous session, you were told which parameters to use for the prior normalinverse-gamma distribution. This time you have to choose the prior parameters yourself.
Given the information below, find reasonable values for the prior parameters of the normal-inverse-gamma distribution — μ₀, ν₀, α₀, β₀. You will be asked to provide your values for the prior hyperparameters in class, and to explain how you came up with them.

Information provided:

● The data are normally distributed. The error margins given below represent 1 standard deviation from the mean of the parameter.

● Constraint: the mean of the data is approximately 2.3 ± 0.5.

● Constraint: the variance of the data is approximately 2.75 ± 1.

● Find μ₀, ν₀, α₀, β₀ hyperparameters for the normal-inverse-gamma prior that match this information.
You can solve this problem in at least two different ways. You have to implement at least
one of these two methods. Ideally, implement both and check that your answers match.
1. Frame the information above as an optimization problem. You should design a
function that is minimized when the constraints above are all satisfied. You can do
this by creating an objective function of the sum of squared residuals of the
constraints. Use SciPy to find the minimum of the objective function.
For example, the mean of the data mean should be 2.3. From Wikipedia , we know
that the expected value of the mean of the normal-inverse-gamma distribution is μ0
so we add the term (μ .3) to the objective function. You should have 4 such 0 ­ 2
2
terms in your function, corresponding to the 4 constraints in the information above.
2. The constraints above give you the mean of the data mean, the standard deviation
of the data mean, the mean of the data variance, and the standard deviation of the
data variance. Solve these 4 equations (one for each of the quantities above)
simultaneously to find the values of the prior parameters.

In [37]:
import numpy as np
from scipy.optimize import minimize
def f(x):
  return (1-x[0])**2 + 100*(x[1] - x[0]**2)**2
x_0 = [0, 0]

result = minimize(f, x_0)
x_final = result.x
print('Started at', x_initial)
print('Ended at', x_final)
print('f(%s) = %.6f' % (x_final, f(x_final)))

Started at [0, 0]
Ended at [0.99999467 0.99998932]
f([0.99999467 0.99998932]) = 0.000000


In [76]:
from scipy.optimize import minimize

x_0 = [1,2,3,4] #mu, beta, alpha, lambda


def NIG_min(x):
  mean = x[0] - 2.3
  std = x[1]/(x[2] - 1) - 2.75
  mean_var = x[1]/(x[2]-1)*x[3] - 0.5
  std_var = (x[1]**2/(((x[2]-1)**2) * (x[2]-2))) - 1
  return mean**2 + std**2 + mean_var**2+ std_var**2

result = minimize(NIG_min, x_0)
result.x

array([ 2.30000031, 23.54687636,  9.56250361,  0.18181832])

I assign some random initial value, and output my prior mu, beta, alpha, lambda as 2.3, 23, 9.6, 0.2. 